File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseTrueForAll.cs
using System;
using System.Collections.Generic;
using System.Linq;

class ListTestcases
{
    void Playground()
    {
        var list = new List<int>();

        var bad = list.All(x => true); // Noncompliant {{The collection-specific "TrueForAll" method should be used instead of the "All" extension}}
        //             ^^^

        var nullableBad = list?.All(x => true); // Noncompliant

        var good = list.TrueForAll(x => true); // Compliant

        List<int> DoWork() => null;
        Func<List<int>, bool> func = l => l.All(x => true); // Noncompliant

        var methodBad = DoWork().All(x => true); // Noncompliant
        var methodGood = DoWork().TrueForAll(x => true); // Compliant

        var nullableMethodBad = DoWork()?.All(x => true); // Noncompliant
        var nullableMethodGood = DoWork()?.TrueForAll(x => true); // Compliant

        var inlineInitialization = new List<int> { 42 }.All(x => true); // Noncompliant

        var imposter = new ContainsAllMethod<int>();
        imposter.All(x => true); // Compliant

        var badList = new BadList<int>();
        badList.All(x => true); // Compliant

        var goodList = new GoodList<int>();
        goodList.All(x => true); // Noncompliant

        var ternary = (true ? list : goodList).All(x => true); // Noncompliant
        var nullCoalesce = (list ?? goodList).All(x => true); // Noncompliant
        var ternaryNullCoalesce = (list ?? (true ? list : goodList)).All(x => true); // Noncompliant

         goodList.Fluent().Fluent().Fluent().Fluent().All(x => true); //Noncompliant
         goodList.Fluent().Fluent().Fluent().Fluent()?.All(x => true); //Noncompliant
         goodList.Fluent().Fluent().Fluent()?.Fluent().All(x => true); //Noncompliant
         goodList.Fluent().Fluent().Fluent()?.Fluent()?.All(x => true); //Noncompliant
         goodList.Fluent().Fluent()?.Fluent().Fluent().All(x => true); //Noncompliant
         goodList.Fluent().Fluent()?.Fluent().Fluent()?.All(x => true); //Noncompliant
         goodList.Fluent().Fluent()?.Fluent()?.Fluent().All(x => true); //Noncompliant
         goodList.Fluent().Fluent()?.Fluent()?.Fluent()?.All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent().Fluent().Fluent().All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent().Fluent().Fluent()?.All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent().Fluent()?.Fluent().All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent().Fluent()?.Fluent()?.All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent()?.Fluent().Fluent().All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent()?.Fluent().Fluent()?.All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent()?.Fluent()?.Fluent().All(x => true); //Noncompliant
         goodList.Fluent()?.Fluent()?.Fluent()?.Fluent()?.All(x => true); //Noncompliant
     //                                                   ^^^

        All<int>(x => true); // Compliant
        AcceptMethod<int>(goodList.All); //FN this is not an invocation, just a member access
    }

    bool All<T>(Func<T, bool> predicate) => true;
    void AcceptMethod<T>(Func<Func<T, bool>, bool> methodThatLooksLikeAll) { }

    class GoodList<T> : List<T>
    {
        public GoodList<T> Fluent() => this;

        void CallAll() =>
            this.All(x => true); // Noncompliant
    }

    class BadList<T> : List<T>
    {
        public bool All(Func<T, bool> predicate) => true;
    }

    class ContainsAllMethod<T>
    {
        public bool All(Func<T, bool> predicate) => true;
    }
}

class ArrayTestcases
{
    void Playground()
    {
        var array = new int[42];

        var bad = array.All(x => true); // Noncompliant
        var nullableBad = array?.All(x => true); // Noncompliant

        var good = Array.TrueForAll(array, x => true); // Compliant

        int[] DoWork() => null;
        Func<int[], bool> func = l => l.All(x => true); // Noncompliant

        var methodBad = DoWork().All(x => true); // Noncompliant
        var methodGood = Array.TrueForAll(DoWork(), x => true); // Compliant

        var nullableMethodBad = DoWork()?.All(x => true); // Noncompliant
        var inlineInitialization = new int[] { 42 }.All(x => true); // Noncompliant

         array.ToArray().ToArray().ToArray().ToArray().All(x => true); //Noncompliant
         array.ToArray().ToArray().ToArray().ToArray()?.All(x => true); //Noncompliant
         array.ToArray().ToArray().ToArray()?.ToArray().All(x => true); //Noncompliant
         array.ToArray().ToArray().ToArray()?.ToArray()?.All(x => true); //Noncompliant
         array.ToArray().ToArray()?.ToArray().ToArray().All(x => true); //Noncompliant
         array.ToArray().ToArray()?.ToArray().ToArray()?.All(x => true); //Noncompliant
         array.ToArray().ToArray()?.ToArray()?.ToArray().All(x => true); //Noncompliant
         array.ToArray().ToArray()?.ToArray()?.ToArray()?.All(x => true); //Noncompliant
         array.ToArray()?.ToArray().ToArray().ToArray().All(x => true); //Noncompliant
         array.ToArray()?.ToArray().ToArray().ToArray()?.All(x => true); //Noncompliant
         array.ToArray()?.ToArray().ToArray()?.ToArray().All(x => true); //Noncompliant
         array.ToArray()?.ToArray().ToArray()?.ToArray()?.All(x => true); //Noncompliant
         array.ToArray()?.ToArray()?.ToArray().ToArray().All(x => true); //Noncompliant
         array.ToArray()?.ToArray()?.ToArray().ToArray()?.All(x => true); //Noncompliant
         array.ToArray()?.ToArray()?.ToArray()?.ToArray().All(x => true); //Noncompliant
         array.ToArray()?.ToArray()?.ToArray()?.ToArray()?.All(x => true); //Noncompliant
    }
}

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/UseTrueForAll.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Reflection

Class ListTestcases
    Private Sub Playground()
        Dim list = New List(Of Integer)()

        Dim bad = list.All(Function(x) True) ' Noncompliant
        Dim nullableBad = list?.All(Function(x) True) ' Noncompliant

        Dim good = list.TrueForAll(Function(x) True) ' Compliant

        Dim DoWork As Func(Of List(Of Integer)) = Function() Nothing
        Dim func As Func(Of List(Of Integer), Boolean) = Function(l) l.All(Function(x) True) ' Noncompliant

        Dim methodBad = DoWork().All(Function(x) True) ' Noncompliant
        Dim methodGood = DoWork().TrueForAll(Function(x) True) ' Compliant

        Dim nullableMethodBad = DoWork()?.All(Function(x) True) ' Noncompliant
        Dim nullableMethodGood = DoWork()?.TrueForAll(Function(x) True) ' Compliant

        Dim inlineInitialization = New List(Of Integer) From {42}.All(Function(x) True) ' Noncompliant

        Dim imposter = New ContainsAllMethod(Of Integer)()
        imposter.All(Function(x) True) ' Compliant

        Dim badList = New BadList(Of Integer)()
        badList.All(Function(x) True) ' Compliant

        Dim goodList = New GoodList(Of Integer)()
        goodList.All(Function(x) True) ' Noncompliant

        Dim ternary As List(Of Integer) = If(True, list, goodList)
        ternary.All(Function(x) True) 'Noncompliant

        goodList.Fluent().Fluent().Fluent().Fluent().All(Function(x) True) 'Noncompliant
        goodList.Fluent().Fluent().Fluent().Fluent()?.All(Function(x) True) 'Noncompliant
        goodList.Fluent().Fluent().Fluent()?.Fluent().All(Function(x) True) 'Noncompliant
        goodList.Fluent().Fluent().Fluent()?.Fluent()?.All(Function(x) True) 'Noncompliant
        goodList.Fluent().Fluent()?.Fluent().Fluent().All(Function(x) True) 'Noncompliant
        goodList.Fluent().Fluent()?.Fluent().Fluent()?.All(Function(x) True) 'Noncompliant
        goodList.Fluent().Fluent()?.Fluent()?.Fluent().All(Function(x) True) 'Noncompliant
        goodList.Fluent().Fluent()?.Fluent()?.Fluent()?.All(Function(x) True) 'Noncompliant
        goodList.Fluent()?.Fluent().Fluent().Fluent().All(Function(x) True) 'Noncompliant
        goodList.Fluent()?.Fluent().Fluent().Fluent()?.All(Function(x) True) 'Noncompliant
        goodList.Fluent()?.Fluent().Fluent()?.Fluent().All(Function(x) True) 'Noncompliant
        goodList.Fluent()?.Fluent().Fluent()?.Fluent()?.All(Function(x) True) 'Noncompliant
        goodList.Fluent()?.Fluent()?.Fluent().Fluent().All(Function(x) True) 'Noncompliant
        goodList.Fluent()?.Fluent()?.Fluent().Fluent()?.All(Function(x) True) 'Noncompliant
        goodList.Fluent()?.Fluent()?.Fluent()?.Fluent().All(Function(x) True) 'Noncompliant
    End Sub

    Class GoodList(Of T)
        Inherits List(Of T)

        Public Function Fluent() As GoodList(Of T)
            Return Me
        End Function
    End Class

    Class BadList(Of T)
        Inherits List(Of T)

        Public Function All(predicate As Func(Of T, Boolean)) As Boolean
            Return True
        End Function
    End Class

    Class ContainsAllMethod(Of T)
        Public Function All(predicate As Func(Of T, Boolean)) As Boolean
            Return True
        End Function
    End Class
End Class

Class ArrayTestcases
    Private Sub Playground()
        Dim array = New Integer(41) {}

        Dim bad = array.All(Function(x) True) ' Noncompliant
        Dim nullableBad = array?.All(Function(x) True) ' Noncompliant

        Dim good = array.TrueForAll(array, Function(x) True) ' Compliant

        Dim func As Func(Of Integer(), Boolean) = Function(l) l.All(Function(x) True) ' Noncompliant

        Dim methodBad = DoWork().All(Function(x) True) ' Noncompliant
        Dim methodGood = array.TrueForAll(DoWork(), Function(x) True) ' Compliant

        Dim nullableMethodBad = DoWork()?.All(Function(x) True) ' Noncompliant
        Dim inlineInitialization = {42}.All(Function(x) True) ' FN `.GetTypeInfo(CollectionInitializer)` returns null

        array.ToArray().ToArray().ToArray().ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray().ToArray().ToArray().ToArray()?.All(Function(x) True) 'Noncompliant
        array.ToArray().ToArray().ToArray()?.ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray().ToArray().ToArray()?.ToArray()?.All(Function(x) True) 'Noncompliant
        array.ToArray().ToArray()?.ToArray().ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray().ToArray()?.ToArray().ToArray()?.All(Function(x) True) 'Noncompliant
        array.ToArray().ToArray()?.ToArray()?.ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray().ToArray()?.ToArray()?.ToArray()?.All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray().ToArray().ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray().ToArray().ToArray()?.All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray().ToArray()?.ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray().ToArray()?.ToArray()?.All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray()?.ToArray().ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray()?.ToArray().ToArray()?.All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray()?.ToArray()?.ToArray().All(Function(x) True) 'Noncompliant
        array.ToArray()?.ToArray()?.ToArray()?.ToArray()?.All(Function(x) True) 'Noncompliant
    End Sub

    Private Function DoWork() As Integer()
        Return Nothing
    End Function
End Class

